docs/examples/files.md

```diff
@@ -0,0 +1,231 @@
+        """
+```
+
+Often, it's the case that you have an abundance of a certain type of data within a `.json` file.
+For example, you might have a list of people:
+
+```json
+[
+    {
+        "name": "John Doe",
+        "age": 30,
+        "email": "john@example.com"
+    },
+    {
+        "name": "Jane Doe",
+        "age": 25,
+        "email": "jane@example.com"
+    }
+]
+```
+
+In this case, you can validate the data against a `List[Person]` model:
+
+```python test="skip"
+from typing import List
+
+from pydantic import BaseModel, EmailStr, PositiveInt, TypeAdapter
+
+
+class Person(BaseModel):
+    name: str
+    age: PositiveInt
+    email: EmailStr
+
+
+type_adapter = TypeAdapter(List[Person])  # (1)!
+
+with open('people.json') as f:
+    json_string = f.read()
+    people = type_adapter.validate_json(json_string)
+    print(people)
+    #> [Person(name='John Doe', age=30, email='john@example.com'), Person(name='Jane Doe', age=25, email='jane@example.com')]
+```
+
+1. We use [`TypeAdapter`][pydantic.type_adapter.TypeAdapter] to validate a list of `Person` objects.
+[`TypeAdapter`][pydantic.type_adapter.TypeAdapter] is a Pydantic construct used to validate data against a single type.
+
+## JSON lines files
+
+Similar to validating a list of objects from a `.json` file, you can validate a list of objects from a `.jsonl` file.
+`.jsonl` files are a sequence of JSON objects separated by newlines.
+
+Consider the following `.jsonl` file:
+
+```json
+{"name": "John Doe", "age": 30, "email": "john@example.com"}
+{"name": "Jane Doe", "age": 25, "email": "jane@example.com"}
+```
+
+We can validate this data with a similar approach to the one we used for `.json` files:
+
+```python test="skip"
+from pydantic import BaseModel, EmailStr, PositiveInt
+
+
+class Person(BaseModel):
+    name: str
+    age: PositiveInt
+    email: EmailStr
+
+
+with open('people.jsonl') as f:
+    people = [Person.model_validate_json(line) for line in f]
+    print(people)
+    #> [Person(name='John Doe', age=30, email='john@example.com'), Person(name='Jane Doe', age=25, email='jane@example.com')]
+```
+
+## CSV files
+
+CSV is one of the most common file formats for storing tabular data.
+To validate data from a CSV file, you can use the `csv` module from the Python standard library to load
+the data and validate it against a Pydantic model.
+
+Consider the following CSV file:
+
+```csv
+name,age,email
+John Doe,30,john@example.com
+Jane Doe,25,jane@example.com
+```
+
+Here's how we validate that data:
+
+```py test="skip"
+import csv
+
+from pydantic import BaseModel, EmailStr, PositiveInt
+
+
+class Person(BaseModel):
+    name: str
+    age: PositiveInt
+    email: EmailStr
+
+
+with open('people.csv') as f:
+    reader = csv.DictReader(f)
+    people = [Person.model_validate(row) for row in reader]
+    print(people)
+    #> [Person(name='John Doe', age=30, email='john@example.com'), Person(name='Jane Doe', age=25, email='jane@example.com')]
```

Comment:
```suggestion
with open('people.csv') as f:
    reader = csv.DictReader(f)
    people = [Person.model_validate(row) for row in reader]
print(people)
#> [Person(name='John Doe', age=30, email='john@example.com'), Person(name='Jane Doe', age=25, email='jane@example.com')]
```
Here I think it's okay to do use the `open` thing this way since csv.DictReader wants a buffer, but I'll note you can wrap the read contents with a `StringIO`. But this maybe makes more sense here since you might be able to rely on the csv reader to have lower peak memory usage than if you were to read the whole file into memory and then parse the CSV.

BUT, I would still encourage closing the file as soon as it is safe to do so, as done in the suggestion above

Reply:
Done, makes sense! Thanks for the tip!